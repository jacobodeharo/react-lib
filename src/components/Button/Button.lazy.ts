const LazyButton = () =>
  import('../Button/Button.styles').then(({ ButtonStyles }) => ({
    default: ButtonStyles,
  }));

export { LazyButton };
