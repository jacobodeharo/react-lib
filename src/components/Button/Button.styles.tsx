import React from 'react';
import { ButtonProps } from './Button.types';
import styled, { css } from 'styled-components';

const getBgColor = (hovered: boolean) =>
  hovered
    ? css`var(--c-button-bg-hover, ${({ theme }) => theme.primaryColor})`
    : css`var(--c-button-bg, ${({ theme }) => theme.primaryColorDark})`;

const getFgColor = (hovered: boolean) =>
  hovered
    ? css`var(--c-button-fg-hover, ${({ theme }) => theme.primaryColorText})`
    : css`var(--c-button-fg, ${({ theme }) => theme.primaryColorLight})`;

const ButtonStyles = styled.button<ButtonProps>`
  align-items: center;
  background-color: ${() => getBgColor(false)};
  border: 0;
  border-radius: 5px;
  color: ${() => getFgColor(false)};
  display: inline-flex;
  justify-content: center;
  min-height: 40px;
  padding: 0 20px;
  transition: color 150ms ease-in-out, background-color 150ms ease-in-out;
  outline: none;

  :hover {
    background-color: ${() => getBgColor(true)};
    color: ${() => getFgColor(true)};
    cursor: pointer;
  }
`;
export { ButtonStyles };
