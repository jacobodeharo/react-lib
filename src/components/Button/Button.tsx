import React from 'react';

import { ButtonProps } from './Button.types';

import './Button.scss';

const Button: React.FC<ButtonProps> = ({ text }) => (
  <button className='button'>{text}</button>
);

export default Button;
