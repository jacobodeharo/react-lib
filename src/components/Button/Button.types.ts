import { ComponentProps } from 'react';
import { Theme } from '../theme/theme.types';

export type ButtonProps = ComponentProps<'button'> & {
  theme?: Theme;
};
