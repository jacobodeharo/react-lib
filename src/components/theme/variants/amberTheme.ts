import { Theme } from '../theme.types';

export const AMBER_THEME: Theme = {
  primaryColorDark: '#FFA000',
  primaryColor: '#FFC107',
  primaryColorLight: '#FFECB3',
  primaryColorText: '#212121',
  accentColor: '#009688',
  primaryTextColor: '#212121',
  secondaryTextColor: '#757575',
  dividerColor: '#BDBDBD',
};
