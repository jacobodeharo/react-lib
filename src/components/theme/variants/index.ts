import { GRAY_THEME } from './grayTheme';
import { AMBER_THEME } from './amberTheme';
import type { Theme } from '../theme.types';

export type ThemeID = 'gray' | 'amber';
export type Themes = {
  [P in ThemeID]: Theme;
};

export const availableThemes: Themes = {
  gray: GRAY_THEME,
  amber: AMBER_THEME,
};
