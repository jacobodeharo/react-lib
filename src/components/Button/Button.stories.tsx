import React from 'react';
import { Button } from './Button';
import { ButtonProps } from './Button.types';
import { Meta, Story } from '@storybook/react/types-6-0';
import { ThemeProvider } from '../theme';
import { availableThemes } from '../theme/variants';

export default {
  title: 'Components/Base/Button',
  component: Button,
  argTypes: {
    label: { control: 'text' },
    themeName: {
      control: {
        type: 'select',
        options: Object.keys(availableThemes),
      },
    },
  },
} as Meta;

type StoryButtonProps = ButtonProps & {
  label: string;
  themeName: string;
};

const Template: Story<StoryButtonProps> = (args) => {
  const { label } = args;
  return (
    <ThemeProvider theme={availableThemes[args.themeName]}>
      <Button>{label}</Button>
    </ThemeProvider>
  );
};

export const Primary = Template.bind({});
Primary.args = {
  label: 'Button',
  themeName: 'light',
};
