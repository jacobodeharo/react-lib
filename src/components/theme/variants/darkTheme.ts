import { Theme } from '../theme.types';

export const DARK_THEME: Theme = {
  cText: '#e5d3d8',
  cBackground: '#39393c',
  cAccent: '#afcce9',
  cAccentBackground: '#4141a2',
  cStroke: '#e5d3d8',
};

export const lightTheme: Theme = {
  ...DARK_THEME,
};
