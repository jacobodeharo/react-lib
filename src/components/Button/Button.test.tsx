import * as React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Button } from './Button';
import { DARK_THEME } from '../theme/variants/darkTheme';
import DoneCallback = jest.DoneCallback;

describe('Button', () => {
  const clickEventHandler = (callback: DoneCallback) => callback();
  const BUTTON_LABEL = 'Button Label';

  afterEach(cleanup);

  test('has to be rendered', () => {
    render(<Button>Button Label</Button>);
    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toMatchSnapshot();
  });

  test('has to accept theming', () => {
    render(<Button theme={DARK_THEME}>Button Label</Button>);
    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toMatchSnapshot();
  });

  test('has to handle onClick events', (done) => {
    render(<Button onClick={clickEventHandler(done)}>{BUTTON_LABEL}</Button>);
    userEvent.click(screen.getByText(BUTTON_LABEL));
  });
});
