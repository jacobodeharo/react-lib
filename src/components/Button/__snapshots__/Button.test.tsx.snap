// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has to accept theming 1`] = `
<button
  class="sc-bdnylx jgaHVl"
>
  Button Label
</button>
`;

exports[`Button has to be rendered 1`] = `
<button
  class="sc-bdnylx Hvhxi"
>
  Button Label
</button>
`;
